idr/query.go

```diff
@@ -0,0 +1,96 @@
+package idr
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/antchfx/xpath"
+	"github.com/jf-tech/go-corelib/caches"
+)
+
+var (
+	// ErrNoMatch is returned when not a single matched node can be found.
+	ErrNoMatch = errors.New("no match")
+	// ErrMoreThanExpected is returned when more than expected matched nodes are found.
+	ErrMoreThanExpected = errors.New("more than expected matched")
+)
+
+const (
+	// DisableXPathCache disables caching xpath compilation when MatchAll/MatchSingle
+	// are called. Useful when caller knows the xpath string isn't cache-able (such as
+	// containing unique IDs, timestamps, etc) which would otherwise cause the xpath
+	// compilation cache grow unbounded.
+	DisableXPathCache = uint(1) << iota
+)
+
+func loadXPathExpr(exprStr string, flags []uint) (*xpath.Expr, error) {
+	var flagsActual uint
+	switch len(flags) {
+	case 0:
+		flagsActual = 0
+	case 1:
+		flagsActual = flags[0]
+	default:
+		return nil, fmt.Errorf("only one flag is allowed, instead got: %d", len(flags))
+	}
+	var expr interface{}
+	var err error
+	if flagsActual&DisableXPathCache != 0 {
+		expr, err = xpath.Compile(exprStr)
+	} else {
+		expr, err = caches.GetXPathExpr(exprStr)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("xpath '%s' compilation failed: %s", exprStr, err.Error())
+	}
+	return expr.(*xpath.Expr), nil
+}
+
+// QueryIter initiates an xpath query specified by 'expr' against an IDR tree rooted at 'n'.
+func QueryIter(n *Node, expr *xpath.Expr) *xpath.NodeIterator {
+	return expr.Select(createNavigator(n))
+}
+
+// AnyMatch returns true if the xpath query 'expr' against an IDR tree rooted at 'n' yields any result.
+func AnyMatch(n *Node, expr *xpath.Expr) bool {
+	return QueryIter(n, expr).MoveNext()
+}
+
+// MatchAll returns all the matched nodes by an xpath query 'exprStr' against an IDR tree rooted at 'n'.
+func MatchAll(n *Node, exprStr string, flags ...uint) ([]*Node, error) {
+	if exprStr == "." {
+		return []*Node{n}, nil
+	}
+	exp, err := loadXPathExpr(exprStr, flags)
+	if err != nil {
+		return nil, err
+	}
+	iter := QueryIter(n, exp)
+	var ret []*Node
+	for iter.MoveNext() {
+		ret = append(ret, nodeFromIter(iter))
+	}
+	return ret, nil
+}
+
+// MatchSingle returns one and only one matched node by an xpath query 'exprStr' against an IDR tree rooted
+// at 'n'. If no matching node is found, ErrNoMatch is returned; if more than one matching nodes are found,
+// ErrMoreThanExpected is returned.
+func MatchSingle(n *Node, exprStr string, flags ...uint) (*Node, error) {
+	if exprStr == "." {
+		return n, nil
+	}
+	expr, err := loadXPathExpr(exprStr, flags)
+	if err != nil {
+		return nil, err
+	}
+	iter := QueryIter(n, expr)
+	if !iter.MoveNext() {
+		return nil, ErrNoMatch
+	}
+	ret := nodeFromIter(iter)
+	if iter.MoveNext() {
+		return nil, ErrMoreThanExpected
+	}
```

Comment:
I guess you are doing twice `MoveNext()` calls instead of invoking `MatchAll` to improve a bit performance, but personally I think may better to add a bit comments here so that reader are a bit more clear.

Reply:
Really? I felt this logic is much easier to understand. If first MoveNext returned false , no match error. If second MoveNext returns true, then multiple match error.

Reply:
Haha, up to you. The first point I was thinking during reading your codes is why not re-use `MatchAll()`

Reply:
Because the inefficiency. Why fail with More Than One Match error only after fetching the entire node set. Doesn't make sense. Microseconds count. :)

Reply:
Actually my previous comment is putting this thinking from PR description to comments. But should be fine, I am okay :-).

Reply:
The reason I don't want to put such comment is:

A comment shouldn't talk about straightforward code it should talk about reasons why we do so.

In this situation, if future reader never know we used MatchAll inside MatchSingle before , they wouldn't even have such curiosity. They would only focus on this double MoveNext logic and to me this logic is self-explanatory warrants no comment.